qiskit/circuit/qpy_serialization.py

```diff
@@ -0,0 +1,706 @@
+import io
+import json
+import struct
+import uuid
+import warnings
+
+import numpy as np
+
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from qiskit.circuit.quantumregister import QuantumRegister
+from qiskit.circuit.classicalregister import ClassicalRegister
+from qiskit.circuit.parameter import Parameter
+from qiskit.circuit.gate import Gate
+from qiskit.circuit.instruction import Instruction
+from qiskit.circuit import library
+from qiskit import circuit as circuit_mod
+from qiskit import extensions
+from qiskit.extensions import quantum_initializer
+from qiskit.version import __version__
+from qiskit.exceptions import QiskitError
+
+# v1 Binary Format
+# ----------------
+# FILE_HEADER
+FILE_HEADER = namedtuple('FILE_HEADER', ['preface', 'qpy_version', 'major_version',
+                                         'minor_version', 'patch_version',
+                                         'num_circuits'])
+FILE_HEADER_PACK = '!6sBBBBQ'
+FILE_HEADER_SIZE = struct.calcsize(FILE_HEADER_PACK)
+
+# HEADER binary format
+HEADER = namedtuple('HEADER', ['name', 'global_phase',
+                               'num_qubits', 'num_clbits', 'metadata_size',
+                               'num_registers', 'num_instructions'])
+HEADER_PACK = '!32sdIIQIQ'
+HEADER_SIZE = struct.calcsize(HEADER_PACK)
+
+# CUSTOM_DEFINITIONS
+# CUSTOM DEFINITION HEADER
+CUSTOM_DEFINITION_HEADER = namedtuple('CUSTOM_DEFINITION_HEADER', ['size'])
+CUSTOM_DEFINITION_HEADER_PACK = '!Q'
+CUSTOM_DEFINITION_HEADER_SIZE = struct.calcsize(CUSTOM_DEFINITION_HEADER_PACK)
+
+# CUSTOM_DEFINITION
+CUSTOM_DEFINITION = namedtuple('CUSTOM_DEFINITON',
+                               ['gate_name', 'type', 'num_qubits'
+                                'num_clbits', 'custom_definition', 'size'])
+CUSTOM_DEFINITION_PACK = '!32s1cII?Q'
+CUSTOM_DEFINITION_SIZE = struct.calcsize(CUSTOM_DEFINITION_PACK)
+
+
+# REGISTER binary format
+REGISTER = namedtuple('REGISTER', ['type', 'size', 'name'])
+REGISTER_PACK = '!1cI10s'
+REGISTER_SIZE = struct.calcsize(REGISTER_PACK)
+
+# INSTRUCTION binary format
+INSTRUCTION = namedtuple('INSTRUCTION', ['name', 'num_parameters', 'num_qargs',
+                                         'num_cargs', 'has_condition',
+                                         'condition_register', 'value'])
+INSTRUCTION_PACK = '!32sHII?10sq'
+INSTRUCTION_SIZE = struct.calcsize(INSTRUCTION_PACK)
+# Instruction argument format
+INSTRUCTION_ARG = namedtuple('INSTRUCTION_ARG', ['type', 'size', 'name'])
+INSTRUCTION_ARG_PACK = '!1cI10s'
+INSTRUCTION_ARG_SIZE = struct.calcsize(INSTRUCTION_ARG_PACK)
+# INSTRUCTION parameter format
+INSTRUCTION_PARAM = namedtuple('INSTRUCTION_PARAM', ['type', 'size'])
+INSTRUCTION_PARAM_PACK = '!1cQ'
+INSTRUCTION_PARAM_SIZE = struct.calcsize(INSTRUCTION_PARAM_PACK)
+# PARAMETER
+PARAMETER = namedtuple('PARAMETER', ['name', 'uuid'])
+PARAMETER_PACK = '!32s16s'
+PARAMETER_SIZE = struct.calcsize(PARAMETER_PACK)
+
+
+def _read_header(file_obj):
```

Comment:
I think it would be better to put all the functions as methods in a `QpyCircuit` class.

Reply:
I'm not sure what advantage this has, these functions are all standalone for the most part and don't really have shared state. At a high level the call stack goes from being something like:

```python3
def dump(self, circuits, file_obj):
    _write_header(file_obj)
    for circuit in circuts:
        _write_circuit(file_obj, circuit)
```
to something like
```python3
class QpyCircuit:

    def __init__(self, file_obj):
        self.file_obj = file_obj

   def dump(self, circuits):
       self.write_header()
       for circuit in circuits:
           self.write_circuits()
````
and `file_obj` (which is arguably the only shared state) isn't actually the same through the whole call stack. For example when we call `write_instructions` as part of `write_circuit` we pass a `BytesIO` buffer and not `file_obj` because we need to write custom instruction definitions prior to the actual circuit instructions. Using a class for all the methods and assuming `file_obj` is an instance attribute makes doing things like that tricky.

Reply:
I see. I was thinking in something closer to `Pickler`. It can be added later.